### ISSUE TITLE
Tow consecutive nn.LayerNorm are used in transformer model when norm_first is False

### ISSUE THREAD

**Issue by csukuangfj** [pytorch/pytorch#74092]
### 🐛 Describe the bug

See the code from the master
https://github.com/pytorch/pytorch/blob/9a5c3ca23ff5f01c78c32ec710ba05942baeb019/torch/nn/modules/transformer.py#L62-L67

https://github.com/pytorch/pytorch/blob/9a5c3ca23ff5f01c78c32ec710ba05942baeb019/torch/nn/modules/transformer.py#L340-L345

When `norm_first` is False, the last layer of the encoder layer is `LayerNorm()`.

The problem is that it always puts a LayerNorm in the encoder, see
https://github.com/pytorch/pytorch/blob/9a5c3ca23ff5f01c78c32ec710ba05942baeb019/torch/nn/modules/transformer.py#L204-L205

The consequence is the output of the last encoder layer is fed into another layernorm, so two consectuive layer norm layers are used here.

---

**Expected behavior**: When norm_first is False, the Encoder should not contain a layer norm layer.

### Versions

The master

**Comment by MattiaSarti** [comment]
I'd like to propose a solution for this issue. There are different possibilities, though - I can think of two of them that are quite different:
- **if a final normalization is always - even when ```norm_first=True``` - supposed to be applied to the output of the <ins>last</ins> encoder/decoder layer:**
  - an additional parameter ```norm_first``` should be required by ```TransformerEncoder```/```TransformerDecoder``` as well to apply the final normalization only when ```norm_first=True```
- **else:**
  - since the final normalization carried out inside ```TransformerEncoder```/```TransformerDecoder``` looks redundant because alredy carried out internally to ```TransformerEncoderLayer```/```TransformerDecoderLayer```, it could simply be removed and ```TransformerEncoder```/```TransformerDecoder``` would not need the initialization parameter ```norm```

The second solution looks more coherent, as users would not expect a final normalization to be carried out at the end of the encoder/decoder anyway even when specifying ```norm_first=True```.

Any feedback?

**Comment by jbschlosser** [comment]
> Any feedback?

One thing I'm unsure about with the proposed backward-compatible solution is how this would work with previously serialized models. By default, LayerNorm has learnable parameters (i.e. `elementwise_affine=True`) that end up in the serialized form of the model. Running `new_model.load_state_dict(old_state_dict)` will fail because of the extra `state_dict` key.

I think to be truly backwards compatible, the fix needs to be opt-in rather than changing the default behavior of Transformer, at least at first.

**Comment by MattiaSarti** [comment]
Thanks for the feedback, @jbschlosser, you are right.

If you are suggesting to go - at least for the moment - for a backward-compatible solution, then how about simply never applying ```self.norm``` inside the ```forward``` method of ```TransformerEncoder```/```TransformerDecoder```?
It should be truly backward-compatible and give the desired behavior, but at the expense of keeping unused ```norm``` attributes for ```TransformerEncoder```/```TransformerDecoder``` that do not take part in the forward and backward propagations at all, and this would mean that when ```TransformerEncoder```/```TransformerDecoder``` are directly instantiated by users ```norm``` won't have any effect.

Do you have more reasonable backward-compatible solutions to suggest?
